alibi/models/pytorch/tests/test_model.py

```diff
@@ -0,0 +1,618 @@
+
+    expected_loss0_val = loss_fns[0](input=y_pred[0], target=y_true[0]).item()
+    expected_loss1_val = loss_fns[1](input=y_pred[1], target=y_true[1]).item()
+    expected_loss_val = loss_weights[0] * expected_loss0_val + loss_weights[1] * expected_loss1_val
+    assert np.allclose(loss_val, expected_loss_val)
+
+
+@pytest.mark.parametrize('dataset', ['classification'], indirect=True)
+@pytest.mark.parametrize('unimodal_model', [
+    {
+        'input_dim': input_dim,
+        'output_dim': output_dim,
+        'compile': True,
+        'optimizer': optimizer_class,
+        'loss': loss_fn,
+        'metrics': metric
+    }
+], indirect=True)
+def test_compute_metrics_unimodal(unimodal_model, dataset):
+    """ Test if the metric computation for a unimodal model matches the expectation. """
+    y_pred = dataset['output']['unimodal']['y_pred']
+    y_true = dataset['output']['unimodal']['y_true']
+
+    with reset_model(unimodal_model):
+        result = unimodal_model.compute_metrics(y_pred=y_pred, y_true=y_true)
+
+    expected_acc = torch.mean((y_true == torch.argmax(y_pred, dim=-1)).float()).item()
+    assert np.allclose(expected_acc, result['accuracy'])
+
+
+@pytest.mark.parametrize('dataset', ['classification'], indirect=True)
+@pytest.mark.parametrize('multimodal_model', [
+    {
+        'input_dim': input_dim,
+        'output_dims': output_dims,
+        'compile': True,
+        'optimizer': optimizer_class,
+        'loss': loss_fns,
+        'loss_weights': loss_weights,
+        'metrics': metrics
+    }
+], indirect=True)
+def test_compute_metrics_multimodal(multimodal_model, dataset):
+    """ Test if the metrics computation for a multimodal model matches the expectation. """
+    y_pred = dataset['output']['multimodal']['y_pred']
+    y_true = dataset['output']['multimodal']['y_true']
+
+    with reset_model(multimodal_model):
+        results = multimodal_model.compute_metrics(y_pred=y_pred, y_true=y_true)
+
+    expected_acc1 = torch.mean((y_true[0] == torch.argmax(y_pred[0], dim=-1)).float()).item()
+    expected_acc2 = torch.mean((y_true[1] == torch.argmax(y_pred[1], dim=-1)).float()).item()
+    assert np.isclose(expected_acc1, results['output_1_accuracy'])
+    assert np.isclose(expected_acc2, results['output_2_accuracy'])
+
+
+@pytest.mark.parametrize('dataset', ['regression'], indirect=True)
+@pytest.mark.parametrize('unimodal_model', [
+    {
+        'input_dim': input_dim,
+        'output_dim': output_dim,
+        'compile': True,
+        'optimizer': optimizer_class,
+        'loss': nn.MSELoss(reduction='mean'),
+    }
+], indirect=True)
+def test_train_step(unimodal_model, dataset):
+    """ Test if the train step return the appropriate statistics. """
+    # copy the model `state_dict` as it will be modified by the `train_step`. Not really necessary now,
+    # but to avoid any errors in the future if it is the case.
+    state_dict_cpy = deepcopy(unimodal_model.state_dict())
```

Comment:
Is the concern here because of fixture sharing between test functions? Could it be done in a more idiomatic way using "yield" fixtures and/or scoping the model fixture at a test-function level instead of module level?

Reply:
yes, the concern is that the model is shared between tests. To be honest, I am not sure how I can re-write it with "yield". Changing the scope to function level will do the job, but what I tried to do is avoid creating a new object again and again for each test. I guess if we decide to move the fixture to function level, then I think there will be no need for the context manager to reset the metrics and the loss. Not entirely sure what's the best approach ... Any suggestions?

Reply:
I think just initializing a model isn't a huge overhead so think function scope makes sense and we have better guarantees for test isolation.